File: ledger/pricing.js
// the repo pricing formula. data-engine/pricing.py and frontend/src/lib/pricing.ts are the
// other copies, and all three are tested against the same cases in pricing/fixtures.json.
// raw numbers in, a price out

const PRICING_VERSION = 'PRICING-2';

const W_STAR = 0.001, W_FORK = 0.01, W_WATCH = 0.05, W_PR = 1.00, W_ISSUE = 1.00;
const ISSUE_DRAG_CAP = 0.60, BASE_LISTING = 5.00, PRICE_FLOOR = 1.00;

// multiplier in [0.70, 1.00] from whole days since the last push. days is floored so it
// matches recency_multiplier() in data-engine/pricing.py. `now` is injectable so the
// fixtures are deterministic
function recencyMultiplier(pushedAt, now = Date.now()) {
    if (!pushedAt) return 1.0;
    const pushed = new Date(pushedAt);
    if (isNaN(pushed.getTime())) return 1.0;
    const days = Math.floor((now - pushed.getTime()) / 86400000);
    if (days <= 30) return 1.0;
    if (days >= 365) return 0.70;
    return 1.0 - 0.30 * (days - 30) / (365 - 30);
}

// open prs and open issues count as ln(1 + n), so a few thousand of either moves the price
// by a few dollars instead of a few thousand
function activity(n) {
    return Math.log1p(Math.max(0, n));
}

function computePrice({ stars = 0, forks = 0, watchers = null, openIssues = 0, openPrs = null, pushedAt = null }, now = Date.now()) {
    stars = stars || 0; forks = forks || 0; openIssues = openIssues || 0;
    if (watchers == null) watchers = stars * 0.03;           // guess for missing data
    if (openPrs == null) {
        const estPrs = openIssues * 0.15;                    // ~15% of open issues are really prs
        openPrs = estPrs;
        openIssues = Math.max(0, openIssues - estPrs);
    }
    const gross = BASE_LISTING + stars * W_STAR + forks * W_FORK + watchers * W_WATCH + activity(openPrs) * W_PR;
    const debt = Math.min(activity(openIssues) * W_ISSUE, ISSUE_DRAG_CAP * gross); // capped so small repos don't go negative
    const price = (gross - debt) * recencyMultiplier(pushedAt, now);
    return Math.round(Math.max(PRICE_FLOOR, price) * 100) / 100;
}

module.exports = {
    PRICING_VERSION, computePrice, recencyMultiplier, activity,
    W_STAR, W_FORK, W_WATCH, W_PR, W_ISSUE, ISSUE_DRAG_CAP, BASE_LISTING, PRICE_FLOOR,
};

File: data-engine/pricing.py
"""the repo pricing formula, the python copy of ledger/pricing.js. both are tested
against the same cases in pricing/fixtures.json. raw numbers in, a price out, no
network, no database.
"""

import math
from datetime import datetime

# weights are in dollars.
W_STAR = 0.001
W_FORK = 0.01
W_WATCH = 0.05
W_PR = 1.00
W_ISSUE = 1.00

ISSUE_DRAG_CAP = 0.60   # issue drag can remove at most 60% of gross
BASE_LISTING = 5.00     # every listing starts here
PRICE_FLOOR = 1.00      # marks never price below this


def recency_multiplier(pushed_at_iso, now=None):
    """multiplier in [0.70, 1.00] from whole days since the last push. days is floored
    so it matches recencyMultiplier() in ledger/pricing.js."""
    if not pushed_at_iso:
        return 1.0
    try:
        pushed = datetime.strptime(pushed_at_iso, "%Y-%m-%dT%H:%M:%SZ")
    except (ValueError, TypeError):
        return 1.0
    if now is None:
        now = datetime.utcnow()
    days = (now - pushed).days   # whole days
    if days <= 30:
        return 1.0
    if days >= 365:
        return 0.70
    return 1.0 - 0.30 * (days - 30) / (365 - 30)


def _round_half_up(x):
    """round to the cent, half away from zero, to match js Math.round(x*100)/100 for
    the non-negative prices we deal with. python's built-in round() is banker's
    rounding, which would disagree with the ledger on exact half-cents."""
    return math.floor(x * 100 + 0.5) / 100


def compute_price(stars, forks, watchers, open_issues, open_prs, pushed_at, now=None):
    """a repo's price from its raw metrics. watchers and open_prs can be None (the
    search endpoint omits them), in which case they're estimated and the hourly
    refresh corrects them. `now` is injectable so the fixtures are deterministic."""
    stars = stars or 0
    forks = forks or 0
    open_issues = open_issues or 0

    if watchers is None:
        watchers = stars * 0.03
    if open_prs is None:
        est_prs = open_issues * 0.15
        open_prs = est_prs
        open_issues = max(0.0, open_issues - est_prs)

    gross = (BASE_LISTING
             + stars * W_STAR
             + forks * W_FORK
             + watchers * W_WATCH
             + open_prs * W_PR)
    debt = min(open_issues * W_ISSUE, ISSUE_DRAG_CAP * gross)
    price = (gross - debt) * recency_multiplier(pushed_at, now)
    return _round_half_up(max(PRICE_FLOOR, price))
